[clangd] Add tweak to override pure virtuals

marcogmaia opened the pull request:
closes https://github.com/clangd/clangd/issues/1037 
closes https://github.com/clangd/clangd/issues/2240

Example:

```c++
class Base {
public:
  virtual void publicMethod() = 0;

protected:
  virtual auto privateMethod() const -> int = 0;
};

// Before:
//                        // cursor here
class Derived : public Base{}^ ;

// After:
class Derived : public Base {
public:
  void publicMethod() override {
    // TODO: Implement this pure virtual method.
    static_assert(false, "Method `publicMethod` is not implemented.");
  }

protected:
  auto privateMethod() const -> int override {
    // TODO: Implement this pure virtual method.
    static_assert(false, "Method `privateMethod` is not implemented.");
  }
};
```

https://github.com/user-attachments/assets/79de40d9-1004-4c2e-8f5c-be1fb074c6de

Commit:
[clangd] Add tweak to add pure virtual overrides

marcogmaia commented:
@HighCommander4 could you please take a look here, or forward it to anyone who can?

zyn0217 commented on clang-tools-extra/clangd/refactor/tweaks/OverridePureVirtuals.cpp:
```diff
@@ -0,0 +1,366 @@
+}
+
+// Helper to get the string spelling of an AccessSpecifier.
+std::string getAccessSpecifierSpelling(AccessSpecifier AS) {
+  switch (AS) {
+  case AS_public:
+    return "public";
+  case AS_protected:
+    return "protected";
+  case AS_private:
+    return "private";
+  case AS_none:
+    // Should not typically occur for class members.
+    return "";
+  }
+  // Unreachable.
+  return "";
+}
```
We have `getAccessSpelling` defined in `clang/include/clang/Basic/Specifiers.h`

marcogmaia replied:
Thanks for pointing it out. I've changed it.

zyn0217 reviewed (commented):
Just a drive-by comment. Thanks for working on it

marcogmaia committed:
Replace `getAccessSpecifierSpelling` with `getAccessSpelling`

zwuis commented on clang-tools-extra/clangd/refactor/tweaks/OverridePureVirtuals.cpp:
```diff
@@ -0,0 +1,349 @@
+            "    static_assert(false, \"Method `{1}` is not implemented.\");\n"
+            "  }\n",
+            Method->getReturnType().getAsString(LangOpts),
+            Method->getNameAsString(), Params,
```
`std::move(Params)`

zwuis commented on clang-tools-extra/clangd/refactor/tweaks/OverridePureVirtuals.cpp:
```diff
@@ -0,0 +1,349 @@
+    ParamsAsString.reserve(Method->parameters().size());
+    std::transform(Method->param_begin(), Method->param_end(),
+                   std::back_inserter(ParamsAsString), GetParamString);
+    const auto Params = llvm::join(ParamsAsString, ", ");
```
Remove `const` so that `Params` is actually moved below.

zwuis reviewed (commented):
Thank you for your work!

Please add a release note in `clang-tools-extra/docs/ReleaseNotes.rst`.

Please read <https://llvm.org/docs/CodingStandards.html> and follow requirements/suggestions in this document such as preferring `llvm::SmallVector` and not using braces on simple single-statement bodies of `if` statements.

marcogmaia committed:
Refactor code and remove const

marcogmaia committed:
Make use of llvm::any_of and more move operations

marcogmaia committed:
Update `clang-tools-extra` release notes

marcogmaia commented on clang-tools-extra/clangd/refactor/tweaks/OverridePureVirtuals.cpp:
```diff
@@ -0,0 +1,349 @@
+  llvm::DenseSet<const CXXMethodDecl *> VisitedCanonicalMethods;
+  // We declare it as a std::function because we are going to call it
+  // recursively.
+  std::function<void(const CXXRecordDecl *)> Collect;
```
Replaced it with a free function.

marcogmaia committed:
Simplify if/for statements

marcogmaia committed:
Replace std::vector in favor of llvm::SmallVector

marcogmaia reviewed (commented):
@zwuis I've applied your review, thank you.

marcogmaia committed:
Make code action unavailable when already overridden

marcogmaia committed:
Minor spelling fixes

marcogmaia committed:
Merge branch 'main' into clangd-add-pure-virtual-override-tweak

marcogmaia commented on clang-tools-extra/clangd/refactor/tweaks/OverridePureVirtuals.cpp:
```diff
@@ -0,0 +1,296 @@
+}
+
+bool hasAbstractBaseAncestor(const clang::CXXRecordDecl *CurrentDecl) {
+  if (!CurrentDecl || !CurrentDecl->getDefinition())
```
Yes, using `assert` now.

marcogmaia committed:
Apply @zwuis review

Addressing unusual cases of method declaration.
- constexpr/consteval specifier
- Ref-qualifier
- trailing return type

Refactored the logic for the method definition grouping

marcogmaia reviewed (commented):
@zwuis, the second round of review was applied.

marcogmaia committed:
fixup! Apply @zwuis review

marcogmaia commented on clang-tools-extra/clangd/refactor/tweaks/OverridePureVirtuals.cpp:
```diff
@@ -0,0 +1,296 @@
+
+    MethodsString +=
+        llvm::formatv(
+            "  {0} {1}({2}) {3}override {{\n"
```
Now we are handling these cases.

We are using the string representation of the method declaration, and then we use the lexer to remove the "virtual" and "= 0"

marcogmaia commented on clang-tools-extra/clangd/refactor/tweaks/OverridePureVirtuals.cpp:
```diff
@@ -0,0 +1,296 @@
+      // Create a replacement to insert the method declarations.
+      // The replacement is at InsertLoc, has length 0 (insertion), and uses
+      // InsertionText.
+      std::string InsertionText = "\n" + MethodsGroupString;
```
Moved, I've also refactored how we are getting the grouped text.

marcogmaia commented on clang-tools-extra/clangd/refactor/tweaks/OverridePureVirtuals.cpp:
```diff
@@ -0,0 +1,296 @@
+    // AppendLoc is the SourceLocation of the closing brace '}' of the class.
+    // The replacement will insert text *before* this closing brace.
+    SourceLocation AppendLoc = CurrentDeclDef->getBraceRange().getEnd();
+    std::string FinalAppendText = NewSectionsToAppendText;
```
I've opted for a move.

marcogmaia committed:
Merge branch 'main' into clangd-add-pure-virtual-override-tweak

zwuis reviewed (commented):
LGTM if HighCommander4 is happy.

HighCommander4 commented:
> LGTM if HighCommander4 is happy.

Apologies, folks, I'm backlogged on reviews and probably won't have time to look at this for a while. If you're looking for a clangd reviewer, perhaps one of our other clangd maintainers, @llvm-beanz or @kadircet, could help with this one.

marcogmaia committed:
Remove unused variable

Some minor improvements:
- Add assert instead of a check that is always false
- Add final on tweak class

marcogmaia committed:
Merge branch 'main' into clangd-add-pure-virtual-override-tweak

HighCommander4 commented:
> If you're looking for a clangd reviewer, perhaps one of our other clangd maintainers, @llvm-beanz or @kadircet, could help with this one.

(I realized you are probably not able to edit the Reviewers field, so I went ahead and added the mentioned folks.)

marcogmaia commented:
Ping

kadircet commented:
sorry I am also having a hard time prioritizing this ATM, especially to provide some feedback on usefulness (when does this trigger, which methods we implement/insert/override, whether we emit one code action or multiple, ...). Can you give a brief description of what's your intended design around those concerns (preferably as comments in clang-tools-extra/clangd/refactor/tweaks/OverridePureVirtuals.cpp). So that it's easier to provide feedback around those calls, rather than understand what the code does.

Apart from that, as long as `::prepare` is `fast` and `non-crashing` (which seems to be), I don't think I have concerns in terms of hard technical blockers.

marcogmaia committed:
Merge branch 'main' into clangd-add-pure-virtual-override-tweak

marcogmaia committed:
Add design explanation for OverridePureVirtuals tweak

marcogmaia commented:
@kadircet I've added a description about the tweak at the beginning of the file as requested.

marcogmaia committed:
Merge branch 'main' into clangd-add-pure-virtual-override-tweak

marcogmaia commented:
Ping

marcogmaia committed:
Merge branch 'main' into clangd-add-pure-virtual-override-tweak

marcogmaia commented:
Ping

marcogmaia committed:
Merge branch 'main' into clangd-add-pure-virtual-override-tweak

llvm-beanz reviewed (approved):
Sorry for the extreme delay in review. I think everyone trying to keep up with clangd reviews has been swamped and pretty over subscribed.

The implementation here looks clean and follows existing conventions in clangd. It is also quite well commented. The test coverage seems solid.

The one concern I have is that there could be quite a few string allocations and reallocations. Generally I prefer to see code building up strings using `SmallString` and `raw_svector_ostream` to build up the string.

That said, much of the code in other clangd tweaks uses std::string pretty aggressively so I think this is probably fine and we can revisit for performance if we encounter issues.

zwuis commented:
Do you need the help to merge this PR?

kadircet commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -70,6 +70,16 @@ Code completion
+- New ``clangd`` code action: "Override pure virtual methods". When invoked on a
+  class definition, this action automatically generates C++ ``override``
+  declarations for all pure virtual methods inherited from its base classes that
+  have not yet been implemented. The generated method stubs include a ``TODO``
```
i think content starting with `The generated method stubs include a ``TODO`` ...` is too much details for release notes, i'd probably drop.

kadircet reviewed (approved):
thanks and sorry for not getting back to this for so long. i don't think i'll get to follow up and looks like others are also busy.

i have some usability concerns like:
- Generating method bodies in class definitions instead of out-of-line in implementation files.
- Providing implementations for all missing pure virtuals, without giving user a control might be noisy at times.
- Not providing implementations for some overriden virtuals is also something i'd miss.

But I guess most of these are improvements we can build up-on here. AFAICT, `::prepare` is both performance and resilient and we got the docs to figure out intent and fix bugs if need be. So thins LGTM as well.

marcogmaia committed:
Merge remote-tracking branch 'origin/main' into clangd-add-pure-virtual-override-tweak

marcogmaia committed:
Apply @kadircet review

marcogmaia commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -70,6 +70,16 @@ Code completion
 Code actions
 ^^^^^^^^^^^^
 
+- New ``clangd`` code action: "Override pure virtual methods". When invoked on a
```
Agreed. Dropped.

marcogmaia commented:
Ping

qinkunbao commented:
Hi @marcogmaia and @kadircet ,

I believe the this PR broke a few buildbots. Would you mind taking a look?

```
[ RUN      ] OverridePureVirtualsTests.MultiAccessSpecifiersOverride
Built preamble of size 211928 for file /clangd-test/TestTU.cpp version null in 0.00 seconds
 #0 0x000064c6a459954e llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/llvm/lib/Support/Unix/Signals.inc:834:13
 #1 0x000064c6a45964a6 llvm::sys::RunSignalHandlers() /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/llvm/lib/Support/Signals.cpp:105:18
 #2 0x000064c6a459aab8 SignalHandler(int, siginfo_t*, void*) /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/llvm/lib/Support/Unix/Signals.inc:426:38
 #3 0x000077593c045250 (/lib/x86_64-linux-gnu/libc.so.6+0x45250)
 #4 0x000077593c0a3f1c pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0xa3f1c)
 #5 0x000077593c04519e raise (/lib/x86_64-linux-gnu/libc.so.6+0x4519e)
 #6 0x000077593c028902 abort (/lib/x86_64-linux-gnu/libc.so.6+0x28902)
 #7 0x000064c6a399fbdc (/home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm_build_ubsan/tools/clang/tools/extra/clangd/unittests/./ClangdTests+0x3377bdc)
 #8 0x000064c6a399eb6e __sanitizer::Die() /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_termination.cpp:52:5
 #9 0x000064c6a39ab079 (/home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm_build_ubsan/tools/clang/tools/extra/clangd/unittests/./ClangdTests+0x3383079)
#10 0x000064c6a3af6673 llvm::DenseMapBase<llvm::DenseMap<clang::AccessSpecifier, unsigned int, llvm::DenseMapInfo<clang::AccessSpecifier, void>, llvm::detail::DenseMapPair<clang::AccessSpecifier, unsigned int>>, clang::AccessSpecifier, unsigned int, llvm::DenseMapInfo<clang::AccessSpecifier, void>, llvm::detail::DenseMapPair<clang::AccessSpecifier, unsigned int>>::initEmpty() /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/llvm/include/llvm/ADT/DenseMap.h:443:18
#11 0x000064c6a3af725d grow /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/llvm/include/llvm/ADT/DenseMap.h:0:64
#12 0x000064c6a3af725d llvm::detail::DenseMapPair<clang::AccessSpecifier, unsigned int>* llvm::DenseMapBase<llvm::DenseMap<clang::AccessSpecifier, unsigned int, llvm::DenseMapInfo<clang::AccessSpecifier, void>, llvm::detail::DenseMapPair<clang::AccessSpecifier, unsigned int>>, clang::AccessSpecifier, unsigned int, llvm::DenseMapInfo<clang::AccessSpecifier, void>, llvm::detail::DenseMapPair<clang::AccessSpecifier, unsigned int>>::InsertIntoBucketImpl<clang::AccessSpecifier>(clang::AccessSpecifier const&, llvm::detail::DenseMapPair<clang::AccessSpecifier, unsigned int>*) /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/llvm/include/llvm/ADT/DenseMap.h:0:0
#13 0x000064c6a3af7032 InsertIntoBucket<const clang::AccessSpecifier &> /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/llvm/include/llvm/ADT/DenseMap.h:596:29
#14 0x000064c6a3af7032 std::__1::pair<llvm::DenseMapIterator<clang::AccessSpecifier, unsigned int, llvm::DenseMapInfo<clang::AccessSpecifier, void>, llvm::detail::DenseMapPair<clang::AccessSpecifier, unsigned int>, false>, bool> llvm::DenseMapBase<llvm::DenseMap<clang::AccessSpecifier, unsigned int, llvm::DenseMapInfo<clang::AccessSpecifier, void>, llvm::detail::DenseMapPair<clang::AccessSpecifier, unsigned int>>, clang::AccessSpecifier, unsigned int, llvm::DenseMapInfo<clang::AccessSpecifier, void>, llvm::detail::DenseMapPair<clang::AccessSpecifier, unsigned int>>::try_emplace<>(clang::AccessSpecifier const&) /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/llvm/include/llvm/ADT/DenseMap.h:292:17
#15 0x000064c6a3af6af3 isHandleInSync /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/llvm/include/llvm/ADT/EpochTracker.h:72:43
#16 0x000064c6a3af6af3 operator-> /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/llvm/include/llvm/ADT/DenseMap.h:1289:5
#17 0x000064c6a3af6af3 llvm::MapVector<clang::AccessSpecifier, llvm::SmallVector<clang::CXXMethodDecl const*, 6u>, llvm::DenseMap<clang::AccessSpecifier, unsigned int, llvm::DenseMapInfo<clang::AccessSpecifier, void>, llvm::detail::DenseMapPair<clang::AccessSpecifier, unsigned int>>, llvm::SmallVector<std::__1::pair<clang::AccessSpecifier, llvm::SmallVector<clang::CXXMethodDecl const*, 6u>>, 0u>>::operator[](clang::AccessSpecifier const&) /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/llvm/include/llvm/ADT/MapVector.h:103:15
#18 0x000064c6a3af40a5 collectMissingPureVirtuals /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/clang-tools-extra/clangd/refactor/tweaks/OverridePureVirtuals.cpp:262:55
#19 0x000064c6a3af40a5 clang::clangd::(anonymous namespace)::OverridePureVirtuals::apply(clang::clangd::Tweak::Selection const&) /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/clang-tools-extra/clangd/refactor/tweaks/OverridePureVirtuals.cpp:298:3
#20 0x000064c6a44bd94e operator() /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/clang-tools-extra/clangd/unittests/tweaks/TweakTesting.cpp:73:33
#21 0x000064c6a44bd94e bool llvm::function_ref<bool (clang::clangd::SelectionTree)>::callback_fn<clang::clangd::(anonymous namespace)::applyTweak(clang::clangd::ParsedAST&, llvm::Annotations::Range, llvm::StringRef, clang::clangd::SymbolIndex const*, llvm::vfs::FileSystem*)::$_0>(long, clang::clangd::SelectionTree) /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/llvm/include/llvm/ADT/STLFunctionalExtras.h:46:12
#22 0x000064c6a67a3091 operator() /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/llvm/include/llvm/ADT/STLFunctionalExtras.h:69:12
#23 0x000064c6a67a3091 clang::clangd::SelectionTree::createEach(clang::ASTContext&, clang::syntax::TokenBuffer const&, unsigned int, unsigned int, llvm::function_ref<bool (clang::clangd::SelectionTree)>) /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/clang-tools-extra/clangd/Selection.cpp:1062:9
#24 0x000064c6a44bc289 has_value /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/libcxx_install_ubsan/include/c++/v1/optional:361:82
#25 0x000064c6a44bc289 operator bool /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/libcxx_install_ubsan/include/c++/v1/optional:825:84
#26 0x000064c6a44bc289 clang::clangd::TweakTest::apply(llvm::StringRef, llvm::StringMap<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, llvm::MallocAllocator>*) const /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/clang-tools-extra/clangd/unittests/tweaks/TweakTesting.cpp:100:8
#27 0x000064c6a44a3f9a Compare<const char *, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, nullptr> /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/third-party/unittest/googletest/include/gtest/gtest.h:1398:12
#28 0x000064c6a44a3f9a clang::clangd::(anonymous namespace)::OverridePureVirtualsTests_MultiAccessSpecifiersOverride_Test::TestBody() /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/clang-tools-extra/clangd/unittests/tweaks/OverridePureVirtualsTests.cpp:392:3
#29 0x000064c6a45c6086 os_stack_trace_getter /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/third-party/unittest/googletest/src/gtest.cc:6240:7
#30 0x000064c6a45c6086 testing::Test::Run() /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/third-party/unittest/googletest/src/gtest.cc:2695:9
#31 0x000064c6a45c7667 os_stack_trace_getter /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/third-party/unittest/googletest/src/gtest.cc:6240:7
#32 0x000064c6a45c7667 testing::TestInfo::Run() /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/third-party/unittest/googletest/src/gtest.cc:2842:11
#33 0x000064c6a45c8b70 testing::TestSuite::Run() /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/third-party/unittest/googletest/src/gtest.cc:3018:9
#34 0x000064c6a45d9937 testing::internal::UnitTestImpl::RunAllTests() /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/third-party/unittest/googletest/src/gtest.cc:5922:15
#35 0x000064c6a45d930f testing::UnitTest::Run() /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/third-party/unittest/googletest/src/gtest.cc:5485:10
#36 0x000064c6a45b0c33 main /home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm-project/third-party/unittest/UnitTestMain/TestMain.cpp:55:3
#37 0x000077593c02a3b8 (/lib/x86_64-linux-gnu/libc.so.6+0x2a3b8)
#38 0x000077593c02a47b __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2a47b)
#39 0x000064c6a3994825 _start (/home/b/sanitizer-x86_64-linux-bootstrap-ubsan/build/llvm_build_ubsan/tools/clang/tools/extra/clangd/unittests/./ClangdTests+0x336c825)
--
exit: -6
```

UBsan
https://lab.llvm.org/buildbot/#/builders/25/builds/10010
Fast
https://lab.llvm.org/buildbot/#/builders/169/builds/13150
